clang/lib/CIR/CodeGen/CIRGenExprAgg.cpp

```diff
@@ -297,62 +301,27 @@ class AggExprEmitter : public StmtVisitor<AggExprEmitter> {
-    CIRGenFunction::SourceLocRAIIObject locRAIIObject{
-        CGF, CGF.getLoc(E->getSourceRange())};
-    // Emit an array containing the elements.  The array is externally
-    // destructed if the std::initializer_list object is.
-    LValue Array = CGF.emitLValue(E->getSubExpr());
-    assert(Array.isSimple() && "initializer_list array not a simple lvalue");
-    Address ArrayPtr = Array.getAddress();
-
-    const ConstantArrayType *ArrayType =
-        Ctx.getAsConstantArrayType(E->getSubExpr()->getType());
-    assert(ArrayType && "std::initializer_list constructed from non-array");
-
-    RecordDecl *Record = E->getType()->castAs<RecordType>()->getDecl();
-    RecordDecl::field_iterator Field = Record->field_begin();
-    assert(Field != Record->field_end() &&
-           Ctx.hasSameType(Field->getType()->getPointeeType(),
-                           ArrayType->getElementType()) &&
-           "Expected std::initializer_list first field to be const E *");
-    // Start pointer.
     auto loc = CGF.getLoc(E->getSourceRange());
-    AggValueSlot Dest = EnsureSlot(loc, E->getType());
-    LValue DestLV = CGF.makeAddrLValue(Dest.getAddress(), E->getType());
-    LValue Start =
-        CGF.emitLValueForFieldInitialization(DestLV, *Field, Field->getName());
-    mlir::Value ArrayStart = ArrayPtr.emitRawPointer();
-    CGF.emitStoreThroughLValue(RValue::get(ArrayStart), Start);
-    ++Field;
-    assert(Field != Record->field_end() &&
-           "Expected std::initializer_list to have two fields");
-
-    auto Builder = CGF.getBuilder();
-
-    auto sizeOp = Builder.getConstInt(loc, ArrayType->getSize());
-
-    mlir::Value Size = sizeOp.getRes();
-    Builder.getUIntNTy(ArrayType->getSizeBitWidth());
-    LValue EndOrLength =
-        CGF.emitLValueForFieldInitialization(DestLV, *Field, Field->getName());
-    if (Ctx.hasSameType(Field->getType(), Ctx.getSizeType())) {
-      // Length.
-      CGF.emitStoreThroughLValue(RValue::get(Size), EndOrLength);
-    } else {
-      // End pointer.
-      assert(Field->getType()->isPointerType() &&
-             Ctx.hasSameType(Field->getType()->getPointeeType(),
-                             ArrayType->getElementType()) &&
-             "Expected std::initializer_list second field to be const E *");
-
-      auto ArrayEnd =
-          Builder.getArrayElement(loc, loc, ArrayPtr.getPointer(),
-                                  ArrayPtr.getElementType(), Size, false);
-      CGF.emitStoreThroughLValue(RValue::get(ArrayEnd), EndOrLength);
+    auto builder = CGF.getBuilder();
+    assert(llvm::isa<MaterializeTemporaryExpr>(E->getSubExpr()));
+    auto *subExpr =
+        llvm::cast<MaterializeTemporaryExpr>(E->getSubExpr())->getSubExpr();
+    assert(llvm::isa<InitListExpr>(subExpr));
+    llvm::SmallVector<mlir::Value> inits{};
+    for (auto *init : llvm::cast<InitListExpr>(subExpr)->inits()) {
+      RValue tmpInit = CGF.emitAnyExprToTemp(init);
+      if (tmpInit.isScalar()) {
```

Comment:
We usually do dispatching in the AST level using `TypeEvaluationKind`, which is also what `emitAnyExprToTemp` already does internally. It does not make sense to dispatch its result `RValue` again here.

Could these lines be replaced by any existing helpers? Probably check out how `CallExpr` gets it done?

Reply:
But here we need `mlir::Value` instead of RValue. CallExpr do the similar things but it contains more complex logic because CallExpr want to construct CallArgList with additional information.
I agree there are definitely some common part but now unfortunately we don't have existing helper function.

Reply:
Thanks for investigating it! Make sense to me :3